app_flutter/lib/state/flutter_build.dart

```diff
@@ -73,69 +64,101 @@ class FlutterBuildState extends ChangeNotifier {
-  // TODO(ianh): make this automatically trigger when listeners are added, and cancel when they're removed.
-  Future<void> startFetchingUpdates() async {
-    if (refreshTimer != null && refreshTimer.isActive) {
-      // There's already an update loop, no need to make another.
-      return;
-    }
+  /// How often to query the Cocoon backend for the current build state.
+  @visibleForTesting
+  final Duration refreshRate = const Duration(seconds: 10);
 
-    /// [Timer.periodic] does not necessarily run at the start of the timer.
-    _fetchBuildStatusUpdate();
+  /// Timer that calls [_fetchStatusUpdates] on a set interval.
+  @visibleForTesting
+  @protected
+  Timer refreshTimer;
 
-    _fetchFlutterBranches().then((List<String> branchResponse) => _branches = branchResponse);
+  // There's no way to cancel futures in the standard library so instead we just track
+  // if we've been disposed, and if so, we drop everything on the floor.
+  bool _active = true;
 
-    refreshTimer = Timer.periodic(refreshRate, (_) => _fetchBuildStatusUpdate());
+  @override
+  void addListener(VoidCallback listener) {
+    if (!hasListeners) {
+      _startFetchingStatusUpdates();
+      assert(refreshTimer != null);
+    }
+    super.addListener(listener);
   }
 
-  /// Update build state to be on [branch] and erase previous branch data.
-  Future<void> updateCurrentBranch(String branch) async {
-    _currentBranch = branch;
-    _moreStatusesExist = true;
-    _isTreeBuilding = null;
-    _statuses = <CommitStatus>[];
-
-    /// Clear previous branch data from the widgets
-    notifyListeners();
+  @override
+  void removeListener(VoidCallback listener) {
+    super.removeListener(listener);
+    if (!hasListeners) {
+      refreshTimer?.cancel();
+      refreshTimer = null;
+    }
+  }
 
-    /// To prevent delays, make an immediate request for dashboard data.
-    _fetchBuildStatusUpdate();
+  /// Start a fixed interval loop that fetches build state updates based on [refreshRate].
+  void _startFetchingStatusUpdates() {
+    assert(refreshTimer == null);
+    _fetchStatusUpdates();
+    refreshTimer = Timer.periodic(refreshRate, _fetchStatusUpdates);
   }
 
   /// Request the latest [statuses] and [isTreeBuilding] from [CocoonService].
   ///
   /// If fetched [statuses] is not on the current branch it will be discarded.
-  Future<void> _fetchBuildStatusUpdate() async {
-    await Future.wait(<Future<void>>[
-      cocoonService.fetchCommitStatuses(branch: _currentBranch).then((CocoonResponse<List<CommitStatus>> response) {
+  Future<void> _fetchStatusUpdates([Timer timer]) async {
+    await Future.wait<void>(<Future<void>>[
+      () async {
+        final CocoonResponse<List<String>> response = await cocoonService.fetchFlutterBranches();
+        if (!_active) {
+          return null;
+        }
+        if (response.error != null) {
+          _errors.send('$errorMessageFetchingBranches: ${response.error}');
```

Comment:
the error brook handles notifying it's listeners, right? (I saw the notifyListener was brought into the else logic block)

Reply:
Yeah, the error brook has a different set of listeners than the state.